### PR TITLE
Phase 5: AB-UPT Surface Branch — Dedicated Surface/Volume Prediction Paths (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AB-UPT (arXiv 2502.09692) shows that separating surface and volume prediction into dedicated branches substantially improves surface metrics. Our current Transolver model predicts all nodes (surface + volume) through a single shared output head. This is a mismatch: surface nodes carry the critical aerodynamic quantities (p_in, p_oodc, p_tan, p_re) but receive no special treatment in decoding.

A simplified AB-UPT variant: after the Transolver backbone produces node features `fx`, we route surface nodes through a dedicated deeper GatedMLP `SurfaceDecoder` before the output projection, while volume nodes pass through the existing head. This gives the model expressivity budget where it matters most — at the surface — without altering the encoder. Paired with `--pressure_first`, the surface decoder applies separately to each pressure-then-velocity split.

Reference: AB-UPT, arXiv 2502.09692 (Abu-Talib et al., 2025) — "Adaptive Branch Universal Physics Transformer".

## Instructions

All changes go in `cfd_tandemfoil/train.py`.

### Step 1 — Add SurfaceDecoder module

Add a `SurfaceDecoder` class (GatedMLP with configurable depth and width) after existing module definitions:

```python
class SurfaceDecoder(nn.Module):
    """Dedicated deeper MLP for surface node decoding (AB-UPT style)."""
    def __init__(self, in_dim, hidden_dim, out_dim, n_layers=3):
        super().__init__()
        layers = []
        d = in_dim
        for i in range(n_layers):
            out = out_dim if i == n_layers - 1 else hidden_dim
            layers.append(nn.Linear(d, out * 2 if i < n_layers - 1 else out))
            d = out
        self.layers = nn.ModuleList(layers)

    def forward(self, x):
        for i, layer in enumerate(self.layers):
            if i < len(self.layers) - 1:
                h = layer(x)
                x, gate = h.chunk(2, dim=-1)
                x = x * torch.sigmoid(gate)  # gated linear unit
            else:
                x = layer(x)
        return x
```

### Step 2 — Integrate into Model

In `Model.__init__`, after defining the output head(s), add:

```python
if self.surface_branch:
    n_hidden = self.n_hidden  # same as backbone hidden dim
    self.surface_decoder_p = SurfaceDecoder(
        n_hidden, surface_decoder_width, p_out_dim, n_layers=surface_decoder_depth
    )
    if not surface_shared_head:
        self.surface_decoder_v = SurfaceDecoder(
            n_hidden, surface_decoder_width, v_out_dim, n_layers=surface_decoder_depth
        )
```

In `Model.forward`, after the last TransolverBlock produces `fx`, and before the output projection:

```python
if self.surface_branch and is_surface is not None:
    surf_mask = is_surface.bool()  # [B, N]
    fx_surf = fx[surf_mask]        # [n_surf, C]
    # For pressure_first: route through surface_decoder_p / surface_decoder_v
    # then write back into fx at surf_mask positions before the output head
    # (or bypass the output head entirely for surface nodes)
```

Use `--surface_branch` flag to enable. Thread `--surface_decoder_depth` (int, default 2), `--surface_decoder_width` (int, default 256), `--surface_shared_head` (store_true, use same decoder for p and v) through `parse_args`.

### Step 3 — Run 8-parallel GPU sweep

Use `--wandb_group phase5_surface_branch` on all runs. Run these 8 configs in parallel:

| Run | surface_decoder_depth | surface_decoder_width | surface_shared_head | notes |
|-----|-----------------------|-----------------------|---------------------|-------|
| 1 | 1 | 256 | no | shallow, wide |
| 2 | 2 | 256 | no | **default — start here** |
| 3 | 3 | 256 | no | deep |
| 4 | 2 | 192 | no | narrow |
| 5 | 2 | 256 | yes | shared p+v head |
| 6 | 3 | 192 | no | deep + narrow |
| 7 | 2 | 256 | no | seed=42 validation |
| 8 | baseline (no surface_branch) | — | — | clean control |

Each run uses the full baseline command plus surface branch flags:

```bash
python train.py --agent fern --wandb_name "fern/surface_branch_d2_w256" \
  --wandb_group phase5_surface_branch \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad \
  --pressure_first --pressure_deep \
  --surface_branch --surface_decoder_depth 2 --surface_decoder_width 256
```

Adjust depth/width/shared flags and `--wandb_name` per row. For the baseline control run, omit `--surface_branch`.

**Key metrics to report:** val/loss, p_in, p_oodc, p_tan, p_re for every run. Note whether surface vs volume MAE breakdown is available — this experiment should primarily win on surface metrics.

## Baseline

**Current best (Phase 4 — PR #1867, Transolver + pressure_first + pressure_deep):**

| Metric | Mean (4 seeds) | Best single |
|--------|---------------|-------------|
| val/loss | 0.401 ± 0.005 | 0.395 |
| p_in | 12.95 ± 0.3 | 12.6 |
| p_oodc | 8.40 ± 0.4 | 8.0 |
| p_tan | 33.8 ± 0.5 | 33.3 |
| p_re | 24.7 ± 0.2 | 24.4 |

W&B group for baseline: search for runs from PR #1867 in the W&B project.

**Reproduce baseline:**
```bash
python train.py --agent <name> --wandb_name "<name>/baseline" \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad \
  --pressure_first --pressure_deep
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: ## Review — AB-UPT Surface Branch

Best config (d=2, w=192): val/loss **0.3994** (baseline 0.4010, -0.4%). p_tan **32.6** (baseline 33.4, -2.4%).

Marginal improvement within noise but directionally consistent. **Request changes — multi-seed validation:**

| GPU | Config | Seed |
|-----|--------|------|
| 0-3 | d=2, w=192 surface branch | 42-45 |
| 4-5 | d=2, w=192 surface branch | 46-47 |
| 6-7 | Baseline (no surface branch) | 42-43 |

If mean surface branch val/loss < 0.401 across 6 seeds, we merge.
